c2: -O produces incorrect C code (in particular with slice function)

I created a small example that is problematic:

```
$include "prelude.sail"

type xlen : Int = 64
type xlen_bytes : Int = 8
type xlenbits = bits(xlen)
type regbits = bits(5)

val zeros : forall 'n, 'n >= 0. atom('n) -> bits('n)
function zeros n = replicate_bits(0b0, n)

/* Architectural state */

register PC : xlenbits
register nextPC : xlenbits

register Xs : vector(32, dec, xlenbits)
/* Getters and setters for X registers (special case for zeros register, x0) */
val rX : regbits -> xlenbits effect {rreg}

function rX(r) =
  match r {
    0b00000 => EXTZ(0x0),
    0b11111 => EXTZ(slice(PC, 0, 32)),
    _ => Xs[unsigned(r)]
  }

val wX : (regbits, xlenbits) -> unit effect {wreg}

function wX(r, v) =
  if r != 0b00000 then {
    match r {
      _ => Xs[unsigned(r)] = v
    }
  }

overload X = {rX, wX}

/* Instruction decode and execute */
enum iop = {RISCV_ADDI}
scattered union ast

val decode : bits(32) -> option(ast) effect pure

val execute : ast -> unit effect {rmem, rreg, wreg}

/* ****************************************************************** */

/* ADDI */

union clause ast = ITYPE : (bits(12), regbits, regbits, iop)

function clause decode imm : bits(12) @ rs1 : regbits @ 0b000 @ rd : regbits @ 0b0010011
  = Some(ITYPE(imm, rs1, rd, RISCV_ADDI))

function clause execute (ITYPE (imm, rs1, rd, RISCV_ADDI)) =
  let rs1_val = X(rs1) in
  let imm_ext : xlenbits = EXTS(imm) in
  let result = rs1_val + imm_ext in
  X(rd) = EXTS(slice(result, 0 , 32))

function clause decode _ = None()
```

When compiled with:
`sail -c2 -O simple.sail`

It produces C code that is not correct:

```
cc -g -I lib_sail/ -I generated_lib/  -c -o generated_lib/out.o generated_lib/out.c
generated_lib/out.c:2127:51: error: use of undeclared identifier 'PC'
    zgaz34 = (safe_rshift(UINT64_MAX, 64 - 32) & (PC >> INT64_C(0)));
```

This should say: `state->PC`

It seems like this issue only occurs when optimizations are turned on. This can also be reproduced by trying 'c2' on the ARM model

I did a bit of debugging (but didn't complete it due to lack of time/unfamiliarity with the code base). My guess is that one node in the AST should say 'Global' rather than 'Name'.

--- a/src/jib/c_backend.ml
+++ b/src/jib/c_backend.ml
@@ -320,21 +320,21 @@ module C_config(Opts : sig val branch_coverage : out_channel option end) : Confi
                   (* We need to check that id's type hasn't changed due to flow typing *)
                   let _, ctyp' = Bindings.find id ctx.locals in
                   if ctyp_equal ctyp ctyp' then
-                    AV_cval (V_id (name id, ctyp), typ)
+                    AV_cval (V_id (name_or_global ctx id, ctyp), typ)
                   else
                     (* id's type changed due to flow typing, so it's
                        really still heap allocated!  *)
                     v
                 with
                   (* Hack: Assuming global letbindings don't change from flow typing... *)
-                  Not_found -> AV_cval (V_id (name id, ctyp), typ)
+                  Not_found -> AV_cval (V_id (name_or_global ctx id, ctyp), typ)
               end
             else
               v
          | Register (_, _, typ) ->
             let ctyp = convert_typ ctx typ in
             if is_stack_ctyp ctyp && not (never_optimize ctyp) then
-              AV_cval (V_id (name id, ctyp), typ)
+              AV_cval (V_id (global id, ctyp), typ)
             else
               v
          | _ -> v

--- a/src/jib/jib_compile.ml
+++ b/src/jib/jib_compile.ml
@@ -182,6 +182,12 @@ module type Config = sig
   val track_throw : bool
 end
 
+let name_or_global ctx id =
+  if Env.is_register id ctx.local_env || IdSet.mem id (Env.get_toplevel_lets ctx.local_env) then
+    global id
+  else
+    name id
+
 module Make(C: Config) = struct
 
 let ctyp_of_typ ctx typ = C.convert_typ ctx typ
@@ -190,12 +196,6 @@ let rec chunkify n xs =
   match Util.take n xs, Util.drop n xs with
   | xs, [] -> [xs]
   | xs, ys -> xs :: chunkify n ys
-
-let name_or_global ctx id =
-  if Env.is_register id ctx.local_env || IdSet.mem id (Env.get_toplevel_lets ctx.local_env) then
-    global id
-  else
-    name id
 
 let coverage_branch_count = ref 0
 

--- a/src/jib/jib_compile.mli
+++ b/src/jib/jib_compile.mli
@@ -134,3 +134,5 @@ end
    convert several Sail language features, these are sail_assert,
    sail_exit, and sail_cons. *)
 val add_special_functions : Env.t -> Env.t
+
+val name_or_global : ctx -> id -> name

